File: android/app/src/main/java/com/noop/ui/HapticPrefs.kt
package com.noop.ui

import android.content.Context

/**
 * Per-event toggles for NOOP's IN-SESSION strap-haptic cues (#1115 offshoot) — the cues that used to fire
 * unconditionally whenever the feature was in use (Breathing pacer, Interval timer, Live Session, workout
 * start/end, double-tap actions, biofeedback session).
 *
 * Every key is DEFAULT-OFF (opt-in). To avoid a silent regression for people already using these features,
 * a one-time migration turns them ON for an EXISTING install (one that has ever synced or paired a strap)
 * and leaves them OFF for a fresh install. The ambient cues (inactivity / smart-alarm / stress / zone
 * coaching / calls / notifications) keep their OWN existing keys and are not duplicated here.
 *
 * Byte-parity twin of the Apple `HapticPrefs` (same key strings + same default-off + same migration rule),
 * so a `.noopbak` restore reads identically on either OS.
 */
object HapticPrefs {
    private const val NAME = "noop_haptics_prefs"
    private fun p(ctx: Context) = ctx.getSharedPreferences(NAME, Context.MODE_PRIVATE)

    // In-session cue keys (Group B). New gates, default OFF.
    // BREATHING also covers the resonance / biofeedback session cues (on Android these run through the
    // Breathe path — there is no separate biofeedback buzz). Double-tap is deliberately NOT here: it's
    // already opt-in via the DoubleTapAction selection (NONE = off), so a second default-off gate would
    // silently kill a configured action.
    const val BREATHING = "haptics.breathing"
    const val INTERVALS = "haptics.intervals"
    const val LIVE_SESSION = "haptics.liveSession"
    const val WORKOUT = "haptics.workout"

    /** All in-session keys, in display order — drives both the migration and the settings section. */
    val inSessionKeys = listOf(BREATHING, INTERVALS, LIVE_SESSION, WORKOUT)

    private const val MIGRATED_V1 = "haptics.migratedV1"

    /** Whether an in-session cue may fire. Runs the one-time preserve-existing migration on first access. */
    fun enabled(ctx: Context, key: String): Boolean {
        migrateIfNeeded(ctx)
        return p(ctx).getBoolean(key, false)
    }

    fun setEnabled(ctx: Context, key: String, value: Boolean) {
        p(ctx).edit().putBoolean(key, value).apply()
    }

    /**
     * One-time preserve-existing migration. An EXISTING install (already ONBOARDED — `noop.onboarded`)
     * fired these cues unconditionally, so we default them ON there; a fresh install leaves them OFF.
     *
     * MUST run at app startup (NoopApplication.onCreate), BEFORE this session's onboarding: the onboarded
     * flag is the "has run a prior version" proxy, and reading it at startup is what keeps a fresh install
     * (not yet onboarded on its first launch) OFF — a lazy read after the user onboards would wrongly see
     * onboarded=true and migrate a brand-new user ON. Idempotent + @Synchronized. Byte-parity twin of the
     * Apple `HapticPrefs.migrateIfNeeded` (same `noop.onboarded` signal).
     */
    @Synchronized
    fun migrateIfNeeded(ctx: Context) {
        val prefs = p(ctx)
        if (prefs.getBoolean(MIGRATED_V1, false)) return
        val establishedInstall = NoopPrefs.of(ctx).getBoolean(NoopPrefs.KEY_ONBOARDED, false)
        prefs.edit().apply {
            if (establishedInstall) inSessionKeys.forEach { putBoolean(it, true) }
            putBoolean(MIGRATED_V1, true)
        }.apply()
    }
}

File: Strand/Data/HapticPrefs.swift
import Foundation

/// Per-event toggles for NOOP's IN-SESSION strap-haptic cues (#1115 offshoot) — the byte-parity twin of
/// the Android `HapticPrefs` (same key strings, same default-off, same preserve-existing migration), so a
/// `.noopbak` restore reads identically on either OS.
///
/// Every key is DEFAULT-OFF (opt-in). A one-time migration turns them ON for an EXISTING install (already
/// onboarded) so an update doesn't silence a feature mid-use, and leaves a fresh install OFF. Ambient cues
/// (inactivity / smart-alarm / stress / zone coaching) keep their OWN existing keys and are not duplicated
/// here. Nonisolated (plain `UserDefaults`) so any actor can read a gate at a buzz site.
enum HapticPrefs {
    // In-session cue keys (Group B). `breathing` also covers the resonance / biofeedback session cues
    // (parity with Android, where those run through the Breathe path). Double-tap is deliberately NOT here:
    // it's already opt-in via the DoubleTapAction picker (a second default-off gate would kill a configured
    // action).
    static let breathing = "haptics.breathing"
    static let intervals = "haptics.intervals"
    static let liveSession = "haptics.liveSession"
    static let workout = "haptics.workout"

    /// All in-session keys, in display order — drives both the migration and the settings section.
    static let inSessionKeys = [breathing, intervals, liveSession, workout]

    private static let migratedV1Key = "haptics.migratedV1"
    private static let onboardedKey = "noop.onboarded"

    /// Whether an in-session cue may fire. Runs the one-time migration as a safety net; the authoritative
    /// run is at app launch (see `migrateIfNeeded`).
    static func enabled(_ key: String, _ d: UserDefaults = .standard) -> Bool {
        migrateIfNeeded(d)
        return d.bool(forKey: key)
    }

    static func setEnabled(_ key: String, _ value: Bool, _ d: UserDefaults = .standard) {
        d.set(value, forKey: key)
    }

    /// One-time preserve-existing migration — twin of Android `HapticPrefs.migrateIfNeeded`. MUST run at
    /// app launch, BEFORE this session's onboarding: `noop.onboarded` is the "has run a prior version"
    /// proxy, and reading it at launch is what keeps a fresh install (not yet onboarded on its first
    /// launch) OFF — a read after the user onboards would wrongly see onboarded=true and migrate a
    /// brand-new user ON. Idempotent.
    static func migrateIfNeeded(_ d: UserDefaults = .standard) {
        if d.bool(forKey: migratedV1Key) { return }
        let establishedInstall = d.bool(forKey: onboardedKey)
        if establishedInstall { for k in inSessionKeys { d.set(true, forKey: k) } }
        d.set(true, forKey: migratedV1Key)
    }
}
